servo/connectors/newrelic.py

```diff
@@ -0,0 +1,340 @@
+from __future__ import annotations
+import asyncio
+from collections import defaultdict
+from dateutil import parser as date_parser
+from functools import reduce
+from datetime import datetime, timedelta
+from typing import Awaitable, Callable, Dict, Iterable, List, Optional, Set, Tuple
+
+import httpx
+import httpcore._exceptions
+from pydantic import BaseModel, AnyHttpUrl, validator
+
+from servo import (
+    BaseChecks,
+    BaseConfiguration,
+    BaseConnector,
+    Control,
+    Check,
+    CheckHandler,
+    Description,
+    Duration,
+    Filter,
+    HaltOnFailed,
+    License,
+    Maturity,
+    Measurement,
+    Metric,
+    Unit,
+    check,
+    metadata,
+    on_event,
+    TimeSeries,
+    DurationProgress,
+)
+from servo.checks import create_checks_from_iterable, multicheck
+from servo.utilities import join_to_series
+
+
+DEFAULT_BASE_URL = 'https://api.newrelic.com'
+API_PATH = "/v2"
+
+# TODO: how are secrets handled?
+NEWRELIC_ACCOUNT_ID = str(open('/run/secrets/optune_newrelic_account_id').read()).strip()
+NEWRELIC_APM_API_KEY = str(open('/run/secrets/optune_newrelic_apm_api_key').read()).strip()
+NEWRELIC_APM_APP_ID = str(open('/run/secrets/optune_newrelic_apm_app_id').read()).strip()
+
+class NewrelicMetric(Metric):
+    """NewrelicMetric objects describe metrics that can be measure from the Newrelic apm api."""    
+    
+    fetch_name: str
+    """The name of the APM metric containing the values for this Metric
```

Comment:
But its not a query. Its not even the complete query since each metric also requires the `values_selector` property unlike prometheus whose metrics could function with the singular `query` property.

I would be open to renaming the properties to `name_param` and `values_param` since they're turned into params for the http request here: https://github.com/opsani/servox/blob/bc8d66cd934ddac09978541bc8d8471a473bf788/servo/connectors/newrelic.py#L140

However, I feel like calling it a query is shoehorning the typing

Reply:
I don't love the `_param` suffix because it's conflating the implementation detail of being shoved into an HTTP request with its core meaning.

Couldn't we just eliminate the notion of the fetch and values selector entirely by adopting the metric naming convention that New Relic uses in their docs (e.g., metric.name=`WebFrontend/Queue:average_response_time`) (https://docs.newrelic.com/docs/apis/rest-api-v2/application-examples-v2/average-response-time-examples-v2):

```
Response time = HttpDispatcher:average_call_time + ((WebFrontend/Queue:call_count * WebFrontend/Queue:average_response_time)/ HttpDispatcher:call_count)
```

What is before the colon becomes the metric name and what is after becomes the values. Throw a regular expression constraint onto the field and it flattens out into a single string.

The downside is that you can't batch up requests against a single metric without doing some preprocessing to group them by the source, but I think you are already in that situation because `values_selector` is just a string.

This is something I tried to handle in my sketch last week by decomposing the access into the notion of a query, which provides the source of the data, and the metric which binds a value from a named query to something the servo can report back to the optimizer (https://gist.github.com/blakewatters/310f0df9dd051d7c3584cccb9d8e6370)

Reply:
See updated implementation: https://github.com/opsani/servox/blob/279ae1fbd2709246baf68615e400c596adf60623/servo/connectors/newrelic.py#L45